Update serial communication and clean up duplicated & outdated info (#1317)
ARMmbed/Handbook: docs/program-setup/serial/serial_communication.md

This allows you to:

- Print out messages to a host PC terminal (useful for debugging).
- Read input from the host PC keyboard.

> - Communicate with applications and programming languages running on the host PC that can communicate with a serial port. Examples are Perl, Python and Java.

Reviewer: Communicate with **programming languages** running on the host PC?
Perhaps replace the whole bullet point with:
`Communicate with applications running on the host computer to exchange data.`

Author: Done

## Hello World - printing messages

This program prints a "Hello World!" message that you can view on a serial terminal. Mbed OS redirects any `printf()` statements to the board's debug USB serial.

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/Tutorials_SerialComm/Serial_STDOUT/)](https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/Tutorials_SerialComm/Serial_STDOUT/main.cpp)

> ### Viewing outputs with `mbed sterm`

Reviewer: ```suggestion
### Viewing output in a serial terminal
```

Serial terminals run on your host PC. They provide an interface where your Mbed board can print and where you can type characters back to your board.

> Mbed CLI has built-in serial support. To open a serial terminal when _one_ Mbed board is connected and using the default baud rate (9600):

Reviewer: ```suggestion
Mbed CLI provides a serial terminal which is configured with a default baud rate of `9600`. When single board is connected, simply run: 
```

Why say that it works only when one board is connected?

Author: As a step-by-step guide, we want to show the simplest scenario first and introduce more options later

```
mbed sterm

> ```

Reviewer: ```
mbed sterm
```
Note: Specify the board communication port with `-p` if more than one board is connected to the host computer.

Author: This is explained later. I'll add a link to point the user to the next section.

Author: Done

> Messages printed after this point will be displayed, so to run the example application again, use the board's reset button or press Ctrl + B on the serial terminal. Now the application restarts and prints "Hello World!" to the terminal.

Reviewer: ```suggestion
Messages printed after this point will be displayed, restart the application using the board's reset button or press `Ctrl + B` on the serial terminal.
The console prints "Hello World!" to the terminal after the reset.
```

To exit the serial terminal, press Ctrl + C.

<span class="tips">**Tip:** To compile, flash and open a serial terminal in one command line, you can append `--flash --sterm` to `mbed compile`.</span>

## Additional options of `mbed sterm`

> More options are supported, run `mbed sterm -h` to show:

Reviewer: ```suggestion
More options are supported, run `mbed sterm -h` to show the help message:
```

    usage: mbed sterm [-h] [-m TARGET] [-p PORT] [-b BAUDRATE] [-e ECHO] [-r] [-v] [-vv]

    Open serial terminal to connected target (usually board), or connect to a user-specified COM port

    optional arguments:
    -h, --help            show this help message and exit
    -m TARGET, --target TARGET
                            Compile target MCU. Example: K64F, NUCLEO_F401RE,
                            NRF51822...
    -p PORT, --port PORT  Communication port. Default: auto-detect. Specifying
                            this will also ignore the -m/--target option above.
    -b BAUDRATE, --baudrate BAUDRATE
                            Communication baudrate. Default: 9600
    -e ECHO, --echo ECHO  Switch local echo on/off. Default: on
    -r, --reset           Reset the targets (via SendBreak) before opening
                            terminal.
    -v, --verbose         Verbose diagnostic output

>     -vv, --very_verbose   Very verbose diagnostic output

Reviewer: Can you put this section between ``` for formatting?

Author: My IDE had issues rendering this particular quote when I used ```, so I thought markdown didn't allow some of the hyphen? I'll try again.

Author: Works now

Author: Done

This allows you to:

- Print out messages to a host PC terminal (useful for debugging).
- Read input from the host PC keyboard.

> - Communicate with applications running on the host computer to exchange data.

Reviewer 2: It needs to be either "host computer" or "host PC" consistently. Although I might just go with "PC" or "computer" - does the word "host" really clarify anything?

Author: It's a leftover from the old documentation, I'll remove it

## Hello World - printing messages

This program prints a "Hello World!" message that you can view on a serial terminal. Mbed OS redirects any `printf()` statements to the board's debug USB serial.

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/Tutorials_SerialComm/Serial_STDOUT/)](https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/Tutorials_SerialComm/Serial_STDOUT/main.cpp)

### Viewing output in a serial terminal

Serial terminals run on your host PC. They provide an interface where your Mbed board can print and where you can type characters back to your board.

> Mbed CLI provides a serial terminal which is configured with a default baud rate of `9600`. When single board is connected, simply run: 

Reviewer: ```suggestion
Mbed CLI provides a serial terminal which is configured with a default baud rate of `9600`. When a single board is connected, simply run: 
```

Author: Cheers

> Mbed CLI provides a serial terminal which is configured with a default baud rate of `9600`. When a single board is connected, simply run: 

Reviewer 2: ```suggestion
Mbed CLI provides a serial terminal that is configured with a default baud rate of `9600`. When a single board is connected, run: 
```

```
mbed sterm
```

(If you have multiple boards connected, please refer to [Additional options of mbed sterm](#additional-options-of-mbed-sterm).)

> Messages printed after this point will be displayed, restart the application using the board's reset button or press `Ctrl + B` on the serial terminal.

Reviewer 2: ```suggestion
`mbed sterm` starts message printing.

Restart the application using the board's reset button, or by entering <kbd>Ctrl + B</kbd> in the serial terminal.
```

> The console prints "Hello World!" to the terminal after the reset.

Reviewer 2: We haven't used the word "console" until now. What does it refer to?

Author: Good catch, it should be "example"

> To exit the serial terminal, press Ctrl + C.

Reviewer 2: ```suggestion
To close the serial terminal, enter <kbd>Ctrl + C</kbd>.
```

> <span class="tips">**Tip:** To compile, flash and open a serial terminal in one command line, you can append `--flash --sterm` to `mbed compile`.</span>

Reviewer 2: This seems a little out of context.

Author: I think it's better to have a separate (short) section for it

## Additional options of mbed sterm

> If you have multiple boards connected, run `mbedls` to find out the port of the board you want to use, then run `mbed sterm` with `-p <PORT>` to specify.

Reviewer 2: ```suggestion
- If you have multiple boards connected:
    1. Run `mbedls` to find the port of the board you want to use.
    1. Run `mbed sterm -p <PORT>`.
```

> If your application uses a baud rate, specify with `-b <BAUDRATE>` when opening the serial terminal.

Reviewer 2: ```suggestion
- If your application uses a baud rate, specify it with `-b <BAUDRATE>` when opening the serial terminal.
```

Reviewer 2: Don't they always use baud rate? Is it just if it's not the default set somewhere? Shouldn't it match the board?

Author: Will fix it

> More options are supported, run `mbed sterm -h` to show the help message:

Reviewer 2: ```suggestion
- To list all options, run `mbed sterm -h`:
```

```
usage: mbed sterm [-h] [-m TARGET] [-p PORT] [-b BAUDRATE] [-e ECHO] [-r] [-v] [-vv]

Open serial terminal to connected target (usually board), or connect to a user-specified COM port

optional arguments:
-h, --help            show this help message and exit
-m TARGET, --target TARGET
                        Compile target MCU. Example: K64F, NUCLEO_F401RE,
                        NRF51822...
-p PORT, --port PORT  Communication port. Default: auto-detect. Specifying
                        this will also ignore the -m/--target option above.
-b BAUDRATE, --baudrate BAUDRATE
                        Communication baudrate. Default: 9600
-e ECHO, --echo ECHO  Switch local echo on/off. Default: on
-r, --reset           Reset the targets (via SendBreak) before opening
                        terminal.
-v, --verbose         Verbose diagnostic output
-vv, --very_verbose   Very verbose diagnostic output
```

## Additional examples - reading user inputs

> In addition to printing messages, Mbed OS applications can also read keyboard inputs from the user. This is achievable with the [BufferedSerial](../apis/bufferedserial.html)/[UnbufferedSerial](../apis/unbufferedserial.html) classes.

Reviewer 2: ```suggestion
In addition to printing messages, Mbed OS applications can also read keyboard inputs from the user. This is achievable with the [BufferedSerial](../apis/bufferedserial.html) and [UnbufferedSerial](../apis/unbufferedserial.html) classes.
```
